File: skewcorss.py
import torch
import numpy as np

import argparse

parser = argparse.ArgumentParser(description="")

parser.add_argument("-J",type = float,default=1,help="")
parser.add_argument("-T",type = float,default=1,help="")
parser.add_argument("-iters",type = int,default=20,help="")
parser.add_argument("-cut", type = int, default=20,help="")
parser.add_argument("-test",action = 'store_true',help="")

args = parser.parse_args()
torch.set_grad_enabled(False)

K = torch.tensor([1/args.T*args.J],dtype= torch.float64)

T = torch.empty(2,2,2,2)
ran = [1,-1]
for i0,s0 in enumerate(ran):
    for i1,s1 in enumerate(ran):
        for i2,s2 in enumerate(ran):
            for i3,s3 in enumerate(ran):
                T[i0,i1,i2,i3]=torch.exp(K*(s0*s1+s1*s2+s2*s3+s3*s0))

if args.test:
    print("test")
    M = torch.tensor([[torch.exp(K),torch.exp(-K)],[torch.exp(-K),torch.exp(K)]])
    D3 = torch.zeros(2,2,2)
    for i in range(2):
        D3[i,i,i]=1

    Ttest = torch.einsum("abc,bd,efd,fg,ghq,jq,ijk,ci->aehk",(D3,M,D3,M,D3,M,D3,M))
    Ttestp = torch.einsum("bd,dg,gq,qb->bdgq",(M,M,M,M))
    from numpy.testing import assert_array_almost_equal,assert_array_equal
    assert_array_almost_equal(T.numpy(),Ttest.numpy(),5)
    assert_array_almost_equal(Ttest.numpy(),Ttestp.numpy(),5)

maxCut = args.cut
D = 2
epsilon=0
lnZ = torch.zeros(1)

for n in range(args.iters):
    maxT = T.max()
    T = T/maxT
    lnZ += 2**(args.iters-n)*torch.log(maxT)

    TypeA = T.view(D**2,D**2)

    TypeB = T.permute(0,3,1,2).contiguous().view(D**2,D**2)

    Ua,Sa,Va = torch.svd(TypeA)
    Ub,Sb,Vb = torch.svd(TypeB)

    cut = min(maxCut,min((Sa>epsilon).sum().item(),(Sb>epsilon).sum().item()))
    #print(cut)
    #print(maxT)

    S2 = torch.matmul(Ub[:,:cut],torch.diag(torch.sqrt(Sb[:cut]))).view(D,D,cut)
    S1 = torch.matmul(torch.diag(torch.sqrt(Sb[:cut])),Vb.t()[:cut,:]).view(cut,D,D)
    S3 = torch.matmul(Ua[:,:cut],torch.diag(torch.sqrt(Sa[:cut]))).view(D,D,cut)
    S4 = torch.matmul(torch.diag(torch.sqrt(Sa[:cut])),Va.t()[:cut,:]).view(cut,D,D)

    T = torch.einsum('abc,cde,fdg,nfb->aegn',(S1,S3,S2,S4))
    D = cut
    #import pdb
    #pdb.set_trace()

trace = torch.einsum("aaaa->",T)

lnZ += torch.log(trace)

print("K:",K.item())
print("lnZ:",lnZ.item()/2**(args.iters+1))

File: skewcorss.jl
using LinearAlgebra
using TensorOperations

function main()
    K = 1.0

    iteration = 20

    maxD = 20

    T = zeros(Float64,2,2,2,2)

    no2s(no::Int64) = -2*no+3 ::Int64
    sumS(s0::Int64,s1::Int64,s2::Int64,s3::Int64) = Float64(no2s(s0)*no2s(s1)+no2s(s1)*no2s(s2)+no2s(s2)*no2s(s3)+no2s(s3)*no2s(s0))

    for s0 = 1:2
        for s1 = 1:2
            for s2 = 1:2
                for s3 = 1:2
                    T[s3,s2,s1,s0] = exp(K*sumS(s3,s2,s1,s0))
                end
            end
        end
    end

    D = 2
    epsilon = 0.0

    checkBigger(M::Float64) = M>=epsilon

    lnZ = 0.0

    for n = 1:iteration
        maxVal = maximum(T)
        T /= maxVal
        lnZ += 2^(iteration-n+1)*log(maxVal)
        TypeA = reshape(T,(D^2,D^2))
        TypeB = reshape(permutedims(T,[1 4 2 3]),(D^2,D^2))
        ua,sa,va = svd(TypeA)
        ub,sb,vb = svd(TypeB)
        cut = min(size(filter(checkBigger,sb))[1],size(filter(checkBigger,sa))[1],maxD)
        S2 = reshape(ub[:,1:cut]*Diagonal(sqrt.(sb[1:cut])),(D,D,cut))
        S1 = reshape(Diagonal(sqrt.(sb[1:cut]))*transpose(vb)[1:cut,:],(cut,D,D))
        S3 = reshape(ua[:,1:cut]*Diagonal(sqrt.(sa[1:cut])),(D,D,cut))
        S4 = reshape(Diagonal(sqrt.(sa[1:cut]))*transpose(va)[1:cut,:],(cut,D,D))

        @tensor T[a,e,g,n] := S1[a,b,c]*S3[c,d,e]*S2[f,d,g]*S4[n,f,b]
        #T = Tp
        D = cut
    end

    trace = 0.0
    for i = 1:D
        trace += T[i,i,i,i]
    end
    lnZ += log(trace)

    println("K:",K)
    println("lnZ/N:",lnZ/(2^(iteration+1)))
end

main()
